### PR TITLE
.Net OpenAI V2 - Text to Image Service - Phase 02

### PR DESCRIPTION
**Pull request by RogerBarreto**
- Updated ImageToText service implementation using OpenAI SDK
- Updated ImageToText service API's parameters order (modelId first) and added modelId as required (OpenAI supports both dall-e-2 and dall-e-3)
- Added support for OpenAIClient breaking glass for Image to Text Service
- Added support for custom/Non-default endpoint for Image to Text Service
- Added missing Extensions (Service Collection + Kernel Builder) for Embeddings and Image to Text modalities
- Added missing UnitTest for Embeddings
- Added UT convering Image to Text.
- Added integration tests for ImageTotext

- Resolve Partially #6916

### CODE REVIEW

File: dotnet/src/IntegrationTestsV2/Connectors/OpenAI/OpenAITextToImageTests.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.TextToImage;
+using SemanticKernel.IntegrationTests.TestSettings;
+using Xunit;
+
+namespace SemanticKernel.IntegrationTests.Connectors.OpenAI;
+public sealed class OpenAITextToImageTests
```
Comment (markwallace-microsoft) [inline dotnet/src/IntegrationTestsV2/Connectors/OpenAI/OpenAITextToImageTests.cs:11]: Add a summary comment

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.cs [hunk 2]
```diff
@@ -28,6 +34,16 @@ namespace Microsoft.SemanticKernel.Connectors.OpenAI;
 /// </summary>
 internal partial class ClientCore
 {
+    /// <summary>
+    /// White space constant.
+    /// </summary>
+    private const string WhiteSpace = " ";
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.cs:40]: What about `SingleSpace` as the name as `WhiteSpace` could be lot's of different characters

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    private static void EnsureSizeIsSupported(GeneratedImageSize targetSize, Uri endpoint, string modelId)
+    {
+        // Only be restrictive to the image size if the endpoint is the default OpenAI v1 endpoint
+        if (endpoint != new Uri(OpenAIV1Endpoint)) { return; }
+
+        var supported = (targetSize == GeneratedImageSize.W256xH256 && modelId == "dall-e-2")
+        || (targetSize == GeneratedImageSize.W512xH512 && modelId == "dall-e-2")
+        || targetSize == GeneratedImageSize.W1024xH1024
+        || (targetSize == GeneratedImageSize.W1024xH1792 && modelId == "dall-e-3")
+        || (targetSize == GeneratedImageSize.W1792xH1024 && modelId == "dall-e-3");
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs:73]: Can we create a dictionary with the collections valid image sizes keyed by model id and then search for the existence?

I think this statement will need refactoring eventually e.g. what happens in dall-e-N doesn't support `GeneratedImageSize.W1024xH1024`?

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        if (!supported)
+        {
+            throw new ArgumentOutOfRangeException($"The requested image size {targetSize} is not supported. Supported sizes are: 256x256 (dalle-2 only), 512x512 (dalle-2 only), 1024x1024, 1024x1792 (dalle-3 only), 1792x1024 (dalle-3 only)");
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs:77]: It would be better if the error message only output the valid sizes for the model id being used.

File: dotnet/src/Connectors/Connectors.OpenAIV2.UnitTests/Utils/MoqExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+using System;
+using Microsoft.Extensions.Logging;
+using Moq;
+
+#pragma warning disable CS8620 // Argument cannot be used for parameter due to differences in the nullability of reference types.
+
+namespace SemanticKernel.Connectors.OpenAI.UnitTests;
+
+internal static class MoqExtensions
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAIV2.UnitTests/Utils/MoqExtensions.cs:11]: Move to test utilities

File: dotnet/src/Connectors/Connectors.OpenAIV2/Extensions/ServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+- Removed unnecessary Validation checks (that are already happening in the service/client constructors)
+- Added openAIClient extension for TextToImage service.
+- Changed parameters order for TextToImage service extension (modelId comes first).
+- Made modelId a required parameter of TextToImage services.
+
+*/
+/// <summary>
+/// Sponsor extensions class for <see cref="IServiceCollection"/>.
+/// </summary>
+public static class ServiceCollectionExtensions
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Extensions/ServiceCollectionExtensions.cs:26]: The name is long, but I agree just from a internal SK developer perspective this can be handy.

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.cs [hunk 7]
```diff
@@ -80,21 +96,32 @@ internal partial class ClientCore
...
         this.Endpoint = endpoint ?? httpClient?.BaseAddress;
         if (this.Endpoint is null)
         {
             Verify.NotNullOrWhiteSpace(apiKey); // For Public OpenAI Endpoint a key must be provided.
             this.Endpoint = new Uri(OpenAIV1Endpoint);
         }
+        else if (string.IsNullOrEmpty(apiKey))
+        {
+            // Avoids an exception from OpenAI Client when a custom endpoint is provided without an API key.
+            apiKey = WhiteSpace;
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.cs:111]: No, I added a comment and created this issue: [Allow ApiKey to be optional in the OpenAIClient constructor](https://github.com/openai/openai-dotnet/issues/90) in their repo to track this.

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs (hunk 3 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs:73]: I'm doing an extra mile on the check here compared to the OpenAI SDK that only informs in the XmlDocs and leave the problem to happen in the Runtime, probably with a Http 400 Badrequest.

The suggested approach will need a for loop to check each item within the key of the modelId, might be easy to add a new validation but I'm not fully sure if the processing cost worth the change, I can drop this check also...

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs:77]: I think the error message needs to say whats the error.
The supported sizes are IMO the extra information.

File: dotnet/src/Connectors/Connectors.OpenAIV2/Services/OpenAITextToImageService.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    /// </summary>
+    /// <param name="modelId">The model to use for image generation.</param>
+    /// <param name="apiKey">OpenAI API key, see https://platform.openai.com/account/api-keys</param>
+    /// <param name="organizationId">OpenAI organization id. This is usually optional unless your account belongs to multiple organizations.</param>
+    /// <param name="endpoint">Non-default endpoint for the OpenAI API.</param>
+    /// <param name="httpClient">Custom <see cref="HttpClient"/> for HTTP requests.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    public OpenAITextToImageService(
+        string modelId,
+        string? apiKey = null,
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Services/OpenAITextToImageService.cs:47]: Yes, if you provide a custom/non-default endpoint, the ApiKey is not required.

File: dotnet/src/Connectors/Connectors.OpenAIV2/Services/OpenAITextToImageService.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        string? organizationId = null,
+        Uri? endpoint = null,
+        HttpClient? httpClient = null,
+        ILoggerFactory? loggerFactory = null)
+    {
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Services/OpenAITextToImageService.cs:52]: Those are already verified in the constructor of the `ClientCore`, don't need the extra check in here.

File: dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAIV2/Core/ClientCore.TextToImage.cs]: Method was removed.

File: dotnet/src/SemanticKernel.Abstractions/AI/TextToImage/ITextToImageService.cs [hunk 10]
```diff
@@ -16,15 +20,15 @@ public interface ITextToImageService : IAIService
...
+    /// <param name="prompt">Image generation prompt</param>
     /// <param name="width">Image width in pixels</param>
     /// <param name="height">Image height in pixels</param>
     /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
     /// <returns>Generated image in base64 format or image URL</returns>
     [Experimental("SKEXP0001")]
     public Task<string> GenerateImageAsync(
-        string description,
+        string prompt,
```
Comment (SergeyMenshykh) [inline dotnet/src/SemanticKernel.Abstractions/AI/TextToImage/ITextToImageService.cs:31]: This might be an unnecessary breaking change if the method is called with parameter names before arguments, such as `GenerateImageAsync(description: "generate an image of....", ...)`. Do we truly need to rename it?

File: dotnet/src/InternalUtilities/test/MoqExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using Microsoft.Extensions.Logging;
+using Moq;
+
+#pragma warning disable CS8620 // Argument cannot be used for parameter due to differences in the nullability of reference types.
+
+internal static class MoqExtensions
```
Comment (SergeyMenshykh) [inline dotnet/src/InternalUtilities/test/MoqExtensions.cs:9]: Wouldn't this extension be applied to all projects, including those that are not related to testing?

Comment (SergeyMenshykh) [reply dotnet/src/InternalUtilities/test/MoqExtensions.cs:9]: Clarified offline and the test extensions are not going to be pulled to non test projects.

**Review by markwallace-microsoft — APPROVED** [review]
